web3_proxy/src/stats/mod.rs

```diff
@@ -318,216 +320,177 @@ impl BufferedRpcQueryStats {
-                // 2. user_id -> Balance
+        // ====================== //
+        //     INITIATE DELTAS    //
+        // ====================== //
+        // Calculate Balance Only (No referrer)
+        let mut sender_available_balance_delta = Decimal::from(-1) * self.sum_credits_used;
+        let sender_used_balance_delta = self.sum_credits_used;
+        let mut sender_bonus_applied;
+        // Calculate Referrer Bonuses
+        let mut referrer_balance_delta = Decimal::from(0);
+
+        // ============================================================ //
+        //  BASED ON REFERRERS, CALCULATE HOW MUCH SHOULD BE ATTRIBUTED //
+        // ============================================================ //
+        // If we don't lock the database as we do above on the referral_entry, we would have to do this operation on the database
+        if let Some((referral_entity, referrer_code_entity)) = referral_objects {
+            sender_bonus_applied = referral_entity.credits_applied_for_referee;
+
+            // Calculate if we are above the usage threshold, and apply a bonus
+            // Optimally we would read this from the balance, but if we do it like this, we only have to lock a single table (much safer w.r.t. deadlocks)
+            if !referral_entity.credits_applied_for_referee
+                && (referral_entity.credits_applied_for_referrer * (Decimal::from(10))
+                    + self.sum_credits_used)
+                    >= Decimal::from(100)
+            {
+                sender_available_balance_delta += Decimal::from(100);
+                sender_bonus_applied = true;
             }
 
-            // Only downgrade the user in local process memory, not elsewhere
+            // Calculate how much the referrer should get, limited to the last 12 months
+            // Apply 10% of the used balance as a bonus if applicable
+            let now = Utc::now();
+            let valid_until = DateTime::<Utc>::from_utc(referral_entity.referral_start_date, Utc)
+                .checked_add_months(Months::new(12))
+                .unwrap();
 
-            // let mut active_downgrade_user = downgrade_user.into_active_model();
-            // active_downgrade_user.user_tier_id = sea_orm::Set(downgrade_user_role.id);
-            // active_downgrade_user.save(db_conn).await?;
-        }
-
-        // Get the referee, and the referrer
-        // (2) Look up the code that this user used. This is the referee table
-        let referee_object = match referee::Entity::find()
-            .filter(referee::Column::UserId.eq(sender_user_id))
-            .one(db_conn)
-            .await?
-        {
-            Some(x) => x,
-            None => {
-                warn!(
-                    "No referral code was found for this user: {:?}",
-                    sender_user_id
-                );
-                return Ok(());
+            if now <= valid_until {
```

Comment:
so lifetime referrals? 

i agree that this can be circumvened, but i would assume just making it a bit tougher will already have large effect. I don't think everyone will do this (the majority of people wouldnt I suppose)